[Feature] Support managed by external controller

mszadkow opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, check our contributor guidelines: https://www.kubeflow.org/docs/about/contributing
2. To know more about Training Operator, check the developer guide:
    https://github.com/kubeflow/training-operator/blob/master/docs/development/developer_guide.md
3. If you want *faster* PR reviews, check how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:
Add support for `ManagedBy` to be able delegate reconciliation process to external controller.


**Which issue(s) this PR fixes** _(optional, in `Fixes #<issue number>, #<issue number>, ...` format, will close the issue(s) when PR gets merged)_:
Fixes #2193 

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/training/) included if any changes are user facing

mimowo commented on pkg/common/util/webhooks.go:
```diff
@@ -0,0 +1,20 @@
+		allErrs = append(allErrs, validation.IsDomainPrefixedPath(fieldPath, manager)...)
+		if len(manager) > v1.MaxManagedByLength {
+			allErrs = append(allErrs, field.TooLongMaxLength(fieldPath, manager, v1.MaxManagedByLength))
+		}
```
When the value is set verify it is one of the two: 'kubeflow.org/training-operator' or 'kueue.x-k8s.io/multikueue'

mszadkow replied:
Do you think we should introduce awareness of another component, namely Kueue?
I mean we didn't do that in JobSet

mimowo replied:
This is the conclusion from the discussion [here](https://github.com/kubeflow/training-operator/issues/2193#issuecomment-2273468168) and during the meeting. I think this makes sense to keep the status safe until there are valid use cases.  We can bring it up to the JobSet maintainers to consider too.

mszadkow replied:
done

mimowo replied:
Right, I think this is what the current implementation does in this `if`: `if !SupportedJobControllers.Has(manager) {`

mimowo commented on pkg/webhooks/xgboost/xgboostjob_webhook.go:
```diff
@@ -71,10 +72,10 @@ func (w *Webhook) ValidateDelete(context.Context, runtime.Object) (admission.War
 
 func validateXGBoostJob(job *trainingoperator.XGBoostJob) field.ErrorList {
 	var allErrs field.ErrorList
-
```
nit: revert formatting change to minimize distractions

mszadkow replied:
good point

mimowo commented on pkg/webhooks/pytorch/pytorchjob_webhook.go:
```diff
@@ -74,10 +75,10 @@ func (w *Webhook) ValidateDelete(context.Context, runtime.Object) (admission.War
 func validatePyTorchJob(job *trainingoperator.PyTorchJob) (admission.Warnings, field.ErrorList) {
 	var allErrs field.ErrorList
 	var warnings admission.Warnings
-
```
nit: revert formatting changes

mimowo commented on pkg/controller.v1/mpi/mpijob_controller_test.go:
```diff
@@ -1066,6 +1067,69 @@ var _ = Describe("MPIJob controller", func() {
+
+			By("Checking if the pods and services aren't created")
+			Consistently(func() bool {
+				errMasterPod := testK8sClient.Get(ctx, launcherKey, launcherPod)
```
Maybe we could make something to provide useful information when the expectation fails. Will this work to provide more useful information when it fails? 
```golang
Consistently(func() {
  g.Expect(testK8sClient.Get(ctx, launcherKey, launcherPod)).To(testing.BeNotFoundError())
  g.Expect(testK8sClient.Get(ctx, worker0Key, workerPod)).To(testing.BeNotFoundError())
  g.Expect(testK8sClient.Get(ctx, launcherKey, launcherSvc)).To(testing.BeNotFoundError())
  g.Expect(testK8sClient.Get(ctx, worker0Key, workerSvc)).To(testing.BeNotFoundError())
}, util.Timeout, util.Interval).Should(gomega.Succeed())
```

mszadkow replied:
The only comment that comes to my mind is like:
"pods and services should be created by external controller (here not existent)"

mimowo replied:
sgtm

mimowo replied:
what about the suggested changes to fail per object, would it behave as expected?

mszadkow replied:
Seems that only Kueue has those omega matchers - https://github.com/kubernetes-sigs/kueue/pull/1938/files#diff-27451991a8cdca4f9071dca2e2d47b9ac04338cc3d745ae4107b86dcef81ef7b

mimowo replied:
If they are Kueue specific I'm ok to defer it for later if needed. The current formulation of the asserts is ok, imo.

mimowo commented on pkg/controller.v1/mpi/mpijob_controller_test.go:
```diff
@@ -1066,6 +1067,69 @@ var _ = Describe("MPIJob controller", func() {
+				return created.Status.StartTime
+			}, testutil.ConsistentDuration, testutil.Interval).Should(BeNil())
+
+			launcherPod := &corev1.Pod{}
+			workerPod := &corev1.Pod{}
+			launcherSvc := &corev1.Service{}
+			workerSvc := &corev1.Service{}
```
They are not used outside the Consistenlty, we can declare them just there.

mimowo commented on pkg/controller.v1/mpi/mpijob_controller_test.go:
```diff
@@ -1066,6 +1067,69 @@ var _ = Describe("MPIJob controller", func() {
+			Eventually(func() []kubeflowv1.JobCondition {
+				Expect(testK8sClient.Get(ctx, jobKey, created)).Should(Succeed())
+				return created.Status.Conditions
+			}, testutil.ConsistentDuration, testutil.Interval).Should(BeComparableTo([]v1.JobCondition(nil)))
```
```suggestion
			}, testutil.Timeout, testutil.Interval).Should(BeComparableTo([]v1.JobCondition(nil)))
```

mimowo commented on pkg/controller.v1/mpi/mpijob_controller_test.go:
```diff
@@ -1066,6 +1067,69 @@ var _ = Describe("MPIJob controller", func() {
+			}, testutil.Timeout, testutil.Interval).Should(Succeed())
+
+			By("Checking created MPIJob still has a nil startTime")
+			Eventually(func() *metav1.Time {
```
```suggestion
			Consistently(func() *metav1.Time {
```

mimowo commented on pkg/controller.v1/mpi/mpijob_controller_test.go:
```diff
@@ -1066,6 +1067,69 @@ var _ = Describe("MPIJob controller", func() {
+			Eventually(func() *metav1.Time {
+				Expect(testK8sClient.Get(ctx, jobKey, created)).Should(Succeed())
+				return created.Status.StartTime
+			}, testutil.Timeout, testutil.Interval).Should(BeNil())
```
```suggestion
			}, testutil.ConsistentDuration, testutil.Interval).Should(BeNil())
```

mimowo commented on pkg/common/util/webhooks.go:
```diff
@@ -15,6 +15,9 @@ func ValidateManagedBy(runPolicy *v1.RunPolicy, allErrs field.ErrorList) field.E
 		if len(manager) > v1.MaxManagedByLength {
 			allErrs = append(allErrs, field.TooLongMaxLength(fieldPath, manager, v1.MaxManagedByLength))
 		}
+		if manager != v1.MultiKueueController || manager != v1.KubeflowJobsController {
+			allErrs = append(allErrs, field.NotSupported(fieldPath, manager, []string{v1.MultiKueueController, v1.KubeflowJobsController}))
+		}
```
nit: declare the list of supported values in var, and check if the value belongs to them, similarly as here: https://github.com/kubernetes/kubernetes/blob/099a88370d017dacf16e67306ebcdec8394fae83/pkg/apis/batch/validation/validation.go#L83C2-L83C46

mszadkow replied:
good one, thanks

andreyvelich reviewed (commented):
Thank you for this effort @mszadkow!
We will review it soon!
/ok-to-test
/assign @kubeflow/wg-training-leads

andreyvelich commented:
/rerun-all

mszadkow commented:
/retest

andreyvelich commented:
@mszadkow @mimowo I think, I have only one question here: https://github.com/kubeflow/training-operator/pull/2203#discussion_r1719940390.

@kubeflow/wg-training-leads Please take a look at this change.

It would be nice if you could also open PR to update the Kubeflow docs with that API change: https://www.kubeflow.org/docs/components/training/user-guides/.

Maybe we can introduce another user-guide on how to modify `managedBy` API and why it is needed.

tenzen-y commented on pkg/apis/kubeflow.org/v1/common_types.go:
```diff
@@ -221,6 +227,21 @@ type RunPolicy struct {
+	// ManagedBy is used to indicate the controller or entity that manages a job.
+	// The value must be either an empty, 'kubeflow.org/training-operator' or
+	// 'kueue.x-k8s.io/multikueue'.
+	// The built-in job controller reconciles a job which don't have this
```
```suggestion
	// The training-operator reconciles a job which doesn't have this
```

mimowo replied:
+1

tenzen-y commented on pkg/apis/kubeflow.org/v1/common_types.go:
```diff
@@ -221,6 +227,21 @@ type RunPolicy struct {
+	// 'kubeflow.org/training-operator', but delegates reconciling the job
+	// with a 'kueue.x-k8s.io/multikueue' to the Kueue.
+	//
+	// The value must be a valid domain-prefixed path (e.g. acme.io/foo) -
+	// all characters before the first "/" must be a valid subdomain as defined
+	// by RFC 1123. All characters trailing the first "/" must be valid HTTP Path
+	// characters as defined by RFC 3986.
```
```suggestion
```

I think that this should be removed since we do not allow users to specify arbitrary values here.

mimowo replied:
+1

tenzen-y commented on pkg/common/util/webhooks.go:
```diff
@@ -0,0 +1,25 @@
+		fieldPath := field.NewPath("spec", "managedBy")
+		allErrs = append(allErrs, validation.IsDomainPrefixedPath(fieldPath, manager)...)
+		if !SupportedJobControllers.Has(manager) {
+			allErrs = append(allErrs, field.NotSupported(fieldPath, manager, sets.List(SupportedJobControllers)))
```
```suggestion
			allErrs = append(allErrs, field.NotSupported(fieldPath, manager, SupportedJobControllers.UnsortedList()))
```

Do we need an ordered list here?

mimowo replied:
I think the main motivation to return such lists ordered in the core k8s is to make sure that the output is deterministic ([example](https://github.com/kubernetes/kubernetes/blob/746f08a8da0a45e8b1a46362822db05cb11ed294/pkg/apis/batch/validation/validation.go#L334)). This is helpful when writing tests.

tenzen-y commented on pkg/controller.v1/mpi/mpijob_controller.go:
```diff
@@ -135,6 +135,11 @@ func (jc *MPIJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
+	if manager := jc.ManagedByExternalController(mpijob.Spec.RunPolicy); manager != nil {
```
Instead of implementing this skip mechanism for every reconciler, could you implement a predictor?

I think we can implement predictors something like `OnJobCreateFunc` for all Jobs here: https://github.com/kubeflow/training-operator/blob/master/pkg/common/util/reconciler.go.

And then, we can specify the predictor here (this is a PyTorch example): https://github.com/kubeflow/training-operator/blob/c990a4ab846776a1f32e8241dee728aa012d23a5/pkg/controller.v1/pytorch/pytorchjob_controller.go#L182.

The Pod predicator is a good example of a predicator implemented in the TrainingOperator.
- https://github.com/kubeflow/training-operator/blob/e9766d171507ad3772cc04c740acb4d1ee4c544c/pkg/common/util/reconciler.go#L49
- https://github.com/kubeflow/training-operator/blob/c990a4ab846776a1f32e8241dee728aa012d23a5/pkg/controller.v1/pytorch/pytorchjob_controller.go#L200-L204

mszadkow replied:
There is only one mechanism, but called in each reconciler.
How this will differ if I have to setup the predicate in every controller?
What is the benefit of using predicate?
We want to reject the job even before the Reconcile?

tenzen-y replied:
> We want to reject the job even before the Reconcile?

Yes, that my suggestion's motivation.
In a large cluster with hundreds or thousands of Jobs, reconciling performance is an important factor.
By using predicates, we can avoid the enqueue Jobs with managedBy to reconcile queue, and then the approach allows us to avoid performance regressions.

mszadkow replied:
I think I need different predicate for each KubeflowJob as the runPolicy is part of the Spec, the same as `onOwnerCreateFunc`.

> Instead of implementing this skip mechanism for every reconciler, could you implement a predictor?

@tenzen-y  This suggests me you thought differently. 
Wdyt now?

mimowo replied:
We had this discussion when developing in the core k8s, we decided to go for a simple skip in reconciler.

One needs to be careful doing this only in preconditions. Consider the following:
1. user creates a `job-1` without `managedBy`, a sync is queued based on name / namespace
2. the `job-1` is deleted and shortly after `job-2` with the same name and namespace is created, but with `managedBy`
3. the queued reconcile executes and we reconcile `job-2` even though we should not

For this scenario we added this test which was reproducing the potential bug: https://github.com/kubernetes/kubernetes/blob/746f08a8da0a45e8b1a46362822db05cb11ed294/test/integration/job/job_test.go#L2209.

Still, we skip in reconciler + preconditions, but maybe this would be premature optimizing it.

tenzen-y replied:
I had great discussions with @mimowo offline.
Throughout the discussions, we agreeded to implement this managedBy verification mechanism in the reconciler instead of the predicator.

If we implement this verification mechanism only in the predicator, there are some potential bugs in the situation where the Jobs are quickly replaced with the same namespacedName Job with managedBy field, as we discussed k/k: https://github.com/kubernetes/kubernetes/pull/123273#discussion_r1496378689

But, we can improve reconcile queue size by implementing this mechanism both in reconciler and predicator.
However, as a first step, we decided not to implement the predicator, and we will collect user feedback related to controller performance.

tenzen-y reviewed (commented):
Additionally, could you implement E2E testing here?

https://github.com/kubeflow/training-operator/tree/e9766d171507ad3772cc04c740acb4d1ee4c544c/sdk/python/test/e2e

mimowo commented on pkg/apis/kubeflow.org/v1/common_types.go:
```diff
@@ -35,6 +35,12 @@ const (
+	// KubeflowJobsController represents the value of the default jobs controller
+	KubeflowJobsController = "kubeflow.org/training-operator"
+
+	// MultiKueueController represents the value of the only allowed external controller except for KubeflowJobsController
```
```suggestion
	// MultiKueueController represents the value of the only allowed external controller except for KubeflowJobsController
```
```suggestion
	// MultiKueueController represents the MultiKueue controller
```
nit: I think it is unnecessarily repetitive to say here that MultiKueue is the only supported.

mszadkow commented:
Hey @tenzen-y @andreyvelich, I have addressed most of the comments.
What left is:
1. e2e tests - I would be very grateful for any help or advice on how to run these tests locally.
From what I see, they are being run in docker `sdk/python/Dockerfile.conformance` - but seems that they need running cluster or something because that's what I see in result:
```
test/e2e/test_e2e_xgboostjob.py:41: in <module>
    TRAINING_CLIENT = TrainingClient(job_kind=constants.XGBOOSTJOB_KIND)
kubeflow/training/api/training_client.py:79: in __init__
    config.load_kube_config(config_file=config_file, context=context)
/usr/local/lib/python3.10/site-packages/kubernetes/config/kube_config.py:819: in load_kube_config
    loader = _get_kube_config_loader(
/usr/local/lib/python3.10/site-packages/kubernetes/config/kube_config.py:776: in _get_kube_config_loader
    raise ConfigException(
E   kubernetes.config.config_exception.ConfigException: Invalid kube-config file. No configuration found.
=========================== short test summary info ============================
ERROR test/e2e/test_e2e_mpijob.py - kubernetes.config.config_exception.Config...
ERROR test/e2e/test_e2e_paddlejob.py - kubernetes.config.config_exception.Con...
ERROR test/e2e/test_e2e_pytorchjob.py - kubernetes.config.config_exception.Co...
ERROR test/e2e/test_e2e_tfjob.py - kubernetes.config.config_exception.ConfigE...
ERROR test/e2e/test_e2e_xgboostjob.py - kubernetes.config.config_exception.Co...
!!!!!!!!!!!!!!!!!!! Interrupted: 5 errors during collection !!!!!!!!!!!!!!!!!!!!
============================== 5 errors in 0.33s ===============================
```
2. Where should I locate the skip in the end: reconcile, predicate or both?
There are some situations that @mimowo mentioned that it would be beneficial to have it in both places.
https://github.com/kubeflow/training-operator/pull/2203#discussion_r1740468823


Would you agree to make those changes in the follow up possibly?

tenzen-y commented:
> 1. e2e tests - I would be very grateful for any help or advice on how to run these tests locally.
>    From what I see, they are being run in docker `sdk/python/Dockerfile.conformance` - but seems that they need running cluster or something because that's what I see in result:

Actually, the conformance container image does not related to our E2E testing.
So, you can reproduce the CI step locally in the following:

1. Create K8s Cluster: https://github.com/kubeflow/training-operator/blob/00eef586956da5697514d159d3ea4fa4a86d3847/.github/workflows/integration-tests.yaml#L74-L79
2. Build and load new container image to K8s cluster: https://github.com/kubeflow/training-operator/blob/00eef586956da5697514d159d3ea4fa4a86d3847/.github/workflows/integration-tests.yaml#L81-L85
3. Deploy the training-operator with new image: https://github.com/kubeflow/training-operator/blob/00eef586956da5697514d159d3ea4fa4a86d3847/.github/workflows/integration-tests.yaml#L87-L94
4. install Python library, and perform tests: https://github.com/kubeflow/training-operator/blob/00eef586956da5697514d159d3ea4fa4a86d3847/.github/workflows/integration-tests.yaml#L96-L99

mimowo commented on docs/api/kubeflow.org_v1_generated.asciidoc:
```diff
@@ -677,6 +677,14 @@ Suspending a Job will reset the StartTime field of the Job.
+The training-operator reconciles a job which doesn't have this
+field at all or the field value is the reserved string
+'kubeflow.org/training-operator', but delegates reconciling the job
+with a 'kueue.x-k8s.io/multikueue' to the Kueue.
```
```suggestion
with 'kueue.x-k8s.io/multikueue' to the Kueue.
```
nit

mimowo replied:
This seems already addressed. Feel free to resolve the comment.

mimowo reviewed (commented):
The main controller logic and documentation LGTM (except for the nit)

I will defer for completing the review to the maintainers, but maybe we could defer the e2e tests (after manual testing) to a follow up if problematic and could risk the deadline for the next release, WDYT @tenzen-y @andreyvelich ?

tenzen-y commented:
> I will defer completing the review to the maintainers, but maybe we could defer the e2e tests (after manual testing) to a follow up if problematic and could risk the deadline for the next release, WDYT @tenzen-y @andreyvelich ?

We can still contain this feature in the next release since the next release team is still assembling now: https://github.com/kubeflow/community/issues/740

So, I would like to implement all E2E testing in this PR since the E2E is responsible for verifying the control plane and SDK.

andreyvelich commented:
@tenzen-y @mszadkow @mimowo Are we planning to include this feature in the 1.8.1 ?
I thought, since it is not a bug fix, we should include it in the 1.9 release.

tenzen-y commented:
> @tenzen-y @mszadkow @mimowo Are we planning to include this feature in the 1.8.1 ? I thought, since it is not a bug fix, we should include it in the 1.9 release.

We should not include this in the patch release since this is not bug fix, is new feature.

mimowo commented:
Yeah, we should be targeting the new release (1.9) I assume.

mszadkow commented:
> > I will defer completing the review to the maintainers, but maybe we could defer the e2e tests (after manual testing) to a follow up if problematic and could risk the deadline for the next release, WDYT @tenzen-y @andreyvelich ?
> 
> We can still contain this feature in the next release since the next release team is still assembling now: [kubeflow/community#740](https://github.com/kubeflow/community/issues/740)
> 
> So, I would like to implement all E2E testing in this PR since the E2E is responsible for verifying the control plane and SDK.

I was able to run existing E2E locally, so that's a progress.
Though they still fail, which I need to understand why before I add my test case.

tenzen-y commented:
> I was able to run existing E2E locally, so that's a progress.
> Though they still fail, which I need to understand why before I add my test case.

If you can share the error messages here, we can help to resolve that.

mszadkow commented:
@tenzen-y @andreyvelich 
I have added the test, please let me know if this form is acceptable.
I can also split the test in 3 separated ones.
@mimowo you might want to have a look.

Also what about putting the skip in predicate?
Should I do it or not - there is some details in question 2 - https://github.com/kubeflow/training-operator/pull/2203#issuecomment-2328707258

mimowo commented:
> @tenzen-y @andreyvelich I have added the test, please let me know if this form is acceptable. I can also split the test in 3 separated ones. @mimowo you might want to have a look.

Thanks! I will take a look, but maybe this requires yet some fix since the "Go generate Test / Test (pull_request) " Job is failing?

> Also what about putting the skip in predicate? Should I do it or not - there is some details in question 2 - [#2203 (comment)](https://github.com/kubeflow/training-operator/pull/2203#issuecomment-2328707258)

We consulted this with @tenzen-y and decided to defer for later, see [comment](https://github.com/kubeflow/training-operator/pull/2203#discussion_r1744288552). We will guide the decision based on the decisions in the core k8s.

tenzen-y commented on manifests/overlays/standalone/kustomization.yaml:
Why do we need these changes?

mszadkow replied:
not sure, probably "make manifests" result, not done manually

tenzen-y commented on pkg/controller.v1/common/job_test.go:
```diff
@@ -219,6 +219,40 @@ func TestPastActiveDeadline(T *testing.T) {
 
+func TestManagedBy(T *testing.T) {
+	cases := map[string]struct {
+		managedBy string
```
```suggestion
		managedBy *string
```

So that we can avoid the custom insertion loginc in the following, could we make this pointer?

```
if tc.managedBy == "" {
	runPolicy.ManagedBy = nil
}
```

mimowo replied:
+1

tenzen-y commented on pkg/controller.v1/common/job_test.go:
```diff
@@ -219,6 +219,40 @@ func TestPastActiveDeadline(T *testing.T) {
+				runPolicy.ManagedBy = nil
+			}
+
+			if got := jobController.ManagedByExternalController(*runPolicy); got != nil && tc.managedBy != *got {
```
Instead of this custom comparison, could we use the go-cmp?

tenzen-y commented on pkg/common/util/webhooks.go:
```diff
@@ -0,0 +1,23 @@
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+var SupportedJobControllers = sets.New(
```
```suggestion
var supportedJobControllers = sets.New(
```

Can we avoid exposing this variable?

mimowo replied:
+1

tenzen-y commented on pkg/util/testutil/constants.go:
```diff
@@ -16,4 +16,5 @@ const (
 
 var (
 	IgnoreJobConditionsTimes = cmpopts.IgnoreFields(kubeflowv1.JobCondition{}, "LastUpdateTime", "LastTransitionTime")
+	MalformedManagedBy       = "other-job-controller"
```
Where do we use this global variable?

tenzen-y commented:
Once you perform following commands on your local, the errors will be fixed: https://github.com/kubeflow/training-operator/blob/126110fd4d76439bd04ca9fdf96bafb7ea3b6910/Makefile#L39-L52

mimowo commented on sdk/python/test/e2e/utils.py:
```diff
@@ -11,6 +11,14 @@
 logging.getLogger().setLevel(logging.INFO)
 
 
+def verify_externally_managed_job_e2e(client: TrainingClient, name: str, namespace: str):
```
I don't think we need this helper function since it is called from only one place / test, and is highly specific.

mszadkow commented:
/retest-required

mimowo commented on manifests/overlays/standalone/kustomization.yaml:
```diff
@@ -2,12 +2,13 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 namespace: kubeflow
 resources:
-  - ../../base
```
I would suggest to revert the formatting changes. If they are done automatically we can have a dedicated PR for them

mszadkow replied:
Reverted, thanks for pointing out

mimowo commented on pkg/common/util/webhooks.go:
```diff
@@ -0,0 +1,23 @@
+func ValidateManagedBy(runPolicy *v1.RunPolicy, allErrs field.ErrorList) field.ErrorList {
+	if runPolicy.ManagedBy != nil {
+		manager := *runPolicy.ManagedBy
+		fieldPath := field.NewPath("spec", "managedBy")
```
nit: move this under the `if`

mimowo commented on pkg/controller.v1/common/job_test.go:
```diff
@@ -219,6 +220,39 @@ func TestPastActiveDeadline(T *testing.T) {
+	cases := map[string]struct {
+		managedBy *string
+	}{
+		"managedBy is empty": {
```
Add a test case when `managedBy` is nil.

mszadkow replied:
added

mimowo commented on pkg/controller.v1/common/job_test.go:
```diff
@@ -219,6 +220,39 @@ func TestPastActiveDeadline(T *testing.T) {
 
+func TestManagedBy(T *testing.T) {
+	cases := map[string]struct {
+		managedBy *string
```
```suggestion
		managedBy *string
		wantResult bool
```
And provide true/false per test case depending if external controller.

mimowo commented on pkg/controller.v1/common/job_test.go:
```diff
@@ -219,6 +220,39 @@ func TestPastActiveDeadline(T *testing.T) {
 	}
 }
 
+func TestManagedBy(T *testing.T) {
```
```suggestion
func TestManagedByExternalController(T *testing.T) {
```

mimowo commented on pkg/controller.v1/mpi/mpijob_controller.go:
```diff
@@ -135,6 +135,11 @@ func (jc *MPIJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 	}
 
+	if manager := jc.ManagedByExternalController(mpijob.Spec.RunPolicy); manager != nil {
+		logger.Info("Skipping MPIJob managed by a different controller", "managed-by", manager)
```
```suggestion
		logger.Info("Skipping MPIJob managed by a custom controller", "managed-by", manager)
```
nit

mimowo commented on pkg/controller.v1/pytorch/pytorchjob_controller.go:
```diff
@@ -132,6 +132,11 @@ func (r *PyTorchJobReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 
+	if manager := r.ManagedByExternalController(pytorchjob.Spec.RunPolicy); manager != nil {
+		logger.Info("Skipping PyTorchJob managed by a different controller", "managed-by", manager)
```
```suggestion
		logger.Info("Skipping PyTorchJob managed by a custom controller", "managed-by", manager)
```

mimowo commented on pkg/controller.v1/paddlepaddle/paddlepaddle_controller.go:
```diff
@@ -131,6 +131,11 @@ func (r *PaddleJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 	}
 
+	if manager := r.ManagedByExternalController(paddlejob.Spec.RunPolicy); manager != nil {
+		logger.Info("Skipping PaddleJob managed by a different controller", "managed-by", manager)
```
```suggestion
		logger.Info("Skipping PaddleJob managed by a custom controller", "managed-by", manager)
```

mimowo commented on pkg/controller.v1/tensorflow/tfjob_controller.go:
```diff
@@ -127,6 +127,11 @@ func (r *TFJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 	}
 
+	if manager := r.ManagedByExternalController(tfjob.Spec.RunPolicy); manager != nil {
+		logger.Info("Skipping TFJob managed by a different controller", "managed-by", manager)
```
```suggestion
		logger.Info("Skipping TFJob managed by a custom controller", "managed-by", manager)
```

mimowo commented on pkg/controller.v1/xgboost/xgboostjob_controller.go:
```diff
@@ -130,6 +130,11 @@ func (r *XGBoostJobReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 
+	if manager := r.ManagedByExternalController(xgboostjob.Spec.RunPolicy); manager != nil {
+		logger.Info("Skipping XGBoostJob managed by a different controller", "managed-by", manager)
```
```suggestion
		logger.Info("Skipping XGBoostJob managed by a custom controller", "managed-by", manager)
```

mimowo commented on pkg/webhooks/paddlepaddle/paddlepaddle_webhook.go:
```diff
@@ -74,6 +75,7 @@ func validatePaddleJob(job *trainingoperator.PaddleJob) field.ErrorList {
 	if errors := apimachineryvalidation.NameIsDNS1035Label(job.Name, false); len(errors) != 0 {
 		allErrs = append(allErrs, field.Invalid(field.NewPath("metadata").Child("name"), job.Name, fmt.Sprintf("should match: %v", strings.Join(errors, ","))))
 	}
+	allErrs = util.ValidateManagedBy(&job.Spec.RunPolicy, allErrs)
```
IIUC we miss validation that the value is immutable in `ValidateUpdate` functions.

mszadkow commented on pkg/common/util/webhooks.go:
```diff
@@ -0,0 +1,36 @@
+	errs := field.ErrorList{}
+	// Validate immutability
+	if oldRunPolicy != nil && newRunPolicy != nil {
+		if *oldRunPolicy.ManagedBy != *newRunPolicy.ManagedBy {
+			oldManager := *newRunPolicy.ManagedBy
+			newManager := *newRunPolicy.ManagedBy
+			fieldPath := field.NewPath("spec", "runPolicy", "managedBy")
+			errs = append(errs, field.Invalid(fieldPath, newManager, fmt.Sprintf("has changed on update from: %s", oldManager)))
+		}
```
thanks for pointing this out, I was looking for such utility

mimowo commented on pkg/controller.v1/common/job.go:
```diff
@@ -455,3 +455,10 @@ func (jc *JobController) CleanupJob(runPolicy *apiv1.RunPolicy, jobStatus apiv1.
 	return CalcPGMinResources(minMember, replicas, jc.PriorityClassLister.Get)
 }
+
+func (jc *JobController) ManagedByExternalController(rp apiv1.RunPolicy) *string {
```
It is probably better to use `rp *apiv1.RunPolicy` to avoid copying the replica, or `managedBy *string` as suggested in https://github.com/kubeflow/training-operator/pull/2203/files#r1756807676

mimowo commented on pkg/common/util/webhooks.go:
```diff
@@ -0,0 +1,36 @@
+func ValidateManagedBy(oldRunPolicy *v1.RunPolicy, newRunPolicy *v1.RunPolicy) field.ErrorList {
+	errs := field.ErrorList{}
+	// Validate immutability
+	if oldRunPolicy != nil && newRunPolicy != nil {
```
WDYT @tenzen-y @andreyvelich ? It would also be more consistent with `Job` and `JobSet` placement of the field.

mszadkow replied:
Except for a fact that MnagedBy will not be in the same place, but in each spec type I see no issues so far.
The change should be pretty smooth

mimowo replied:
Ah, I see, I got confused, because the code here checks if `runPolicy` is nil, where it cannot be nil, as this is a non-pointer field in spec. 

In that case, we can keep it in spec, and keep the validation code simple (without the runPolicy checks). 

To achieve that I think we need to have two validation functions: one for a new Job (fired on create and update), and validate Job update (fired only on update).

mszadkow replied:
Ok, we found the solution. Closing

mimowo commented on pkg/webhooks/pytorch/pytorchjob_webhook_test.go:
```diff
@@ -284,11 +290,56 @@ func TestValidateV1PyTorchJob(t *testing.T) {
+					trainingoperator.KubeflowJobsController))),
+			},
+		},
+		"managedBy field becomes mutable": {
```
```suggestion
		"attempt to update the managedBy field gets rejected": {
```
same for other tests

mimowo commented on pkg/webhooks/pytorch/pytorchjob_webhook_test.go:
```diff
@@ -284,11 +290,56 @@ func TestValidateV1PyTorchJob(t *testing.T) {
 					specPath.Child("elasticPolicy").Child("nProcPerNode"), specPath.Child("nprocPerNode")),
 			},
 		},
+		"unsupported managedBy controller name": {
```
```suggestion
		"attempt to set unsupported managedBy controller name gets rejected": {
```
same for other tests

mimowo commented on pkg/common/util/webhooks.go:
```diff
@@ -12,22 +12,21 @@ var supportedJobControllers = sets.New(
 	v1.KubeflowJobsController)
 
-func ValidateManagedBy(oldRunPolicy *v1.RunPolicy, newRunPolicy *v1.RunPolicy) field.ErrorList {
+func ValidateRunPolicyCreate(runPolicy *v1.RunPolicy) field.ErrorList {
```
Call it `ValidateRunPolicy` and call it both on `Create` and `Update`.

mimowo commented on pkg/webhooks/pytorch/pytorchjob_webhook.go:
```diff
@@ -55,30 +56,36 @@ func (w *Webhook) ValidateCreate(ctx context.Context, obj runtime.Object) (admis
+	if oldJob != nil {
+		allErrs = append(allErrs, util.ValidateRunPolicyUpdate(&oldJob.Spec.RunPolicy, &newJob.Spec.RunPolicy)...)
+	} else {
+		allErrs = append(allErrs, util.ValidateRunPolicyCreate(&newJob.Spec.RunPolicy)...)
```
This should be called also on Update, to verify the `new`. Same for other jobs.

mimowo commented on pkg/webhooks/paddlepaddle/paddlepaddle_webhook.go:
```diff
@@ -55,26 +56,33 @@ func (w Webhook) ValidateCreate(ctx context.Context, obj runtime.Object) (admiss
+	if oldJob != nil {
+		allErrs = append(allErrs, util.ValidateRunPolicyUpdate(&oldJob.Spec.RunPolicy, &newJob.Spec.RunPolicy)...)
+	} else {
+		allErrs = append(allErrs, util.ValidateRunPolicy(&newJob.Spec.RunPolicy)...)
```
This should be checked unconditionally. We want to run `ValidateRunPolicy` also in case of updates. This may not be needed for immutable fields, but in general this is a good practice in the core k8s.

mimowo commented on pkg/controller.v1/common/job_test.go:
```diff
@@ -219,6 +220,51 @@ func TestPastActiveDeadline(T *testing.T) {
+func TestManagedByExternalController(T *testing.T) {
+	cases := map[string]struct {
+		managedBy  *string
+		wantResult bool
```
```suggestion
		wantResult *string
```

mimowo commented on pkg/controller.v1/common/job_test.go:
```diff
@@ -219,6 +220,51 @@ func TestPastActiveDeadline(T *testing.T) {
+				ManagedBy: tc.managedBy,
+			}
+
+			if got := jobController.ManagedByExternalController(runPolicy.ManagedBy); got != nil {
+				if !tc.wantResult {
+					t.Errorf("Unwanted manager controller: %s\n", *got)
+				}
+				if diff := cmp.Diff(tc.managedBy, got); diff != "" {
+					t.Errorf("Unexpected manager controller (-want +got):\n%s", diff)
+				}
+			}
```
```suggestion
  gotControllerName := jobController.ManagedByExternalController(runPolicy.ManagedBy)
  if cmp.Diff(tc.wantControllerName, gotControllerName) != "" {
    t.Errorf("Unexpected manager controller (-want +got):\n%s", diff)
  }
```
should be more readable, I think

mimowo reviewed (commented):
/lgtm
Reviewed implementation and tests. 
PTAL @tenzen-y and @andreyvelich

tenzen-y reviewed (commented):
@mszadkow Thank you for driving this feature!
@mimowo Additionally, I would like to express my appreciation to help review this PR!

/lgtm
/approve
